Add use_beachmat() helper function

PeteHaitch opened the pull request:
I initially wrote this for my personal R package, but something like it seems a useful addition to **beachmat**.

Basically automates the steps described in the 'Linking' vignette (I added a comment to that vignette).

When run from the directory of a new package (here called **myNewPackage**):

```R
beachmat::use_beachmat()
#> ✔ Adding 'Rcpp' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rhdf5lib' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'beachmat' to LinkingTo field in DESCRIPTION
#> ✔ Adding 'Rcpp' to Imports field in DESCRIPTION*
#> ✔ Adding 'beachmat' to Suggests field in DESCRIPTION*
#> ✔ Adding 'HDF5Array' to Suggests field in DESCRIPTION
#> ✔ Adding 'DelayedArray' to Suggests field in DESCRIPTION
#> ✔ Adding 'Matrix' to Suggests field in DESCRIPTION
#> ✔ Setting SystemRequirements field in DESCRIPTION to 'C++11'
#> ● Include the following directives in your NAMESPACE
#> Copying code to clipboard:
#>   useDynLib('myNewPackage', .registration = TRUE)
#>   importFrom('Rcpp', 'sourceCpp')
#> ● Modify 'NAMESPACE'
#> ● Run document()
```

and finally opens `NAMESPACE` in an text editor so the user can make the required modification.

`beachmat::use_beachmat()` is modelled on `usethis::use_rcpp()` and so:

- Adds dependency on [**usethis**](http://usethis.r-lib.org/)
- Adds a `NOTE` to `R CMD check` because uses some internal functions of **usethis**
- Does sensible things if some of these steps have already been done (e.g., if package already uses **Rcpp**)
- Uses snake_case rather than camelCase (I'm not fussed either way)

BTW I haven't bumped version number.

PeteHaitch committed:
Add use_beachmat() helper function

LTLA commented on DESCRIPTION:
```diff
@@ -4,12 +4,13 @@ Date: 2018-04-25
 Imports: utils, methods, Rhdf5lib (>= 1.1.4), rhdf5, 
         HDF5Array (>= 1.7.3), DelayedArray (>= 0.5.30), 
-        Rcpp (>= 0.12.14) 
+        Rcpp (>= 0.12.14), usethis
```
Can this go into `Suggests:`?

PeteHaitch replied:
Done

LTLA commented on R/use_beachmat.R:
```diff
@@ -0,0 +1,37 @@
+#' \code{vignette("linking", package = "beachmat")}.
+#' @importFrom usethis use_directory use_git_ignore
+#' @export
+use_beachmat <- function() {
```
Four space indenting is the norm in this package.

PeteHaitch replied:
Fixed

LTLA commented on R/use_beachmat.R:
```diff
@@ -0,0 +1,37 @@
+#' For more details see the 'Linking' vignette, available by running
+#' \code{vignette("linking", package = "beachmat")}.
+#' @importFrom usethis use_directory use_git_ignore
+#' @export
```
I'd prefer this not to be exported; it doesn't need to be run in dependent packages every time. If you must export it; I'm not using the Roxygen documentation tools, so you'll have to actually write a manpage manually, i.e., just delete the stuff here and remove the "Do not edit by hand" comments in the manpage.

PeteHaitch replied:
No longer exported although manpage is retained (manual).

PeteHaitch replied:
Also no longer importing anything from **usethis** into `NAMESPACE`

LTLA commented on man/use_beachmat.Rd:
```diff
@@ -0,0 +1,15 @@
+}
+\details{
+For more details see the 'Linking' vignette, available by running
+\code{vignette("linking", package = "beachmat")}.
```
No manual line wrap, one line = one sentence here.

PeteHaitch replied:
Done

LTLA commented on vignettes/linking.Rmd:
````diff
@@ -15,6 +15,11 @@ require(knitr)
 opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
 ```
 
+# Quick setup
````
Move this after or inside the Introduction. And no manual line breaks.

PeteHaitch replied:
Done

PeteHaitch committed:
Changes requested by Aaron in review

LTLA commented on man/use_beachmat.Rd:
```diff
@@ -10,6 +8,5 @@ use_beachmat()
-For more details see the 'Linking' vignette, available by running
-\code{vignette("linking", package = "beachmat")}.
+For more details see the 'Linking' vignette, available by running \code{vignette("linking", package = "beachmat")}.
 }
```
CHECK (or BiocCheck) complains if there's no `\value` or no `\examples`. Also put yourself as `\author`.

LTLA replied:
On hindsight, I suggest deleting this manpage entirely, as it adds absolutely nothing.

LTLA commented on vignettes/linking.Rmd:
```diff
@@ -28,6 +23,10 @@ Currently, the API supports double-precision, integer, logical and character mat
 
+# Quick setup
+
+Run `beachmat:::use_beachmat()` in your package's directory. This will automate most of the steps described below and walk you through the remaining ones.
```
Newline upon sentence start.

PeteHaitch committed:
Changes requested by Aaron in review.

- Add unit tests, based on tests for usethis::use_rcpp()
- Automatically generate Makevars[.win] if not already present

LTLA commented on tests/testthat/helper.R:
```diff
@@ -0,0 +1,50 @@
+proj <- new.env(parent = emptyenv())
```
Can't this entire file go into `test-use_beachmat.R`? It's not like any other function needs it.

PeteHaitch replied:
Yeah, i think so
